### ISSUE TITLE
Long Term Sugestion: Implement Metal API over Vulkan

### ISSUE THREAD

**Issue by Arucard1983** [darlinghq/darling#390]
Beginning with macOS 10-14 Mojave, OpenGL and OpenCL frameworks will be deprecated, and favoured the Apple's Metal API.
Even Apple don't want to support Vulkan API natively, a former proprietary library called MoltenVK translated Vulkan to Metal.

Reversing the MoltenVK logic, it is feasible to implement a Metal to Vulkan compatibility layer ?
Without such support, once Darling starting to run big applications like Adobe Suite or Final Cut that use Metal, would be several underperformed, since native OpenGL support on macOS was poor and outdated.

**Comment by bugaevc** [comment]
> Reversing the MoltenVK logic, it is feasible to implement a Metal to Vulkan compatibility layer ?

Yep, that's the plan. It'd be awesome if someone more familiar with Vulkan and Metal contributed the wrapper 😃

**Comment by CAMR0** [comment]
Ooh. So like DXVK but for metal instead?

**Comment by maddanio** [comment]
that reminds me of a sec talk of some guy that ended up doing dns over http over dns. you should then try metal to vulkan to metal :)

**Comment by hamishmb** [comment]
I believe there's been some work on this with something called Iridium, is that right?

**Comment by totchi-lagawi** [comment]
It's called [Indium](https://github.com/darlinghq/indium/), though it includes a project called Iridium which can translate Metal shaders to SPIR-V (Vulkan) shaders.

**Comment by steelbrain** [comment]
Hey Team! I'm working on my own air2spir-v pipeline and have gotten really far (the goal is to have exact-match output for every single metallib in the macOS system + common apps). I'm minting the goldens with a macbook and verifying them on a linux machine with an NVIDIA GPU through a macOS VM (custom kext that turns para virt GPU commands -> NVIDIA / Vulkan). This is all reverse engineering (AI-assisted). I'm verifying it on a dataset of some 40k Metal Air captures from my machine and its passing a lot of tests and improving a lot every day.

Is there any interest from the darling project in such an effort? I'd love to open source it and share it around if it helps.

**Comment by CuriousTommy** [comment]
> I'm minting the goldens with a macbook and verifying them on a linux machine with an NVIDIA GPU through a macOS VM (custom kext that turns para virt GPU commands -> NVIDIA / Vulkan).

Interesting... is the motivation for this project to create an open source equivalent of Apple's Para-virtual GPU?

> This is all reverse engineering (AI-assisted).

I guess my main concern is if this involves disassembling Apple proprietary code.

> Is there any interest from the darling project in such an effort? I'd love to open source it and share it around if it helps.

Even it we don't end up using this for the Darling project, I still think it would be worthwhile to open source.

**Comment by steelbrain** [comment]
> Interesting... is the motivation for this project to create an open source equivalent of Apple's Para-virtual GPU?

Yes, Vulkan-backed Apple Para-virtual GPU kext.

> I guess my main concern is if this involves disassembling Apple proprietary code.

I'm not a lawyer, and this is not legal advice: Apple hardware and Apple OS is being used to mint goldens that are then output-matched on another device. Based on what I understand from Google vs Oracle, this should be fine?

**Comment by stuaxo** [comment]
Can you explain "minting the goldens" ?

I think I know what you are getting at, I haven't heard the term golden before.

**Comment by steelbrain** [comment]
Hi @stuaxo!

By minting the goldens on a macbook, I meant generating the expected/correct computational output from a trusted machine/GPU. Its for validating the output generated from Metal AIR on a linux machine for example, match = linux is correct, mismatch = linux is wrong.

**Comment by stuaxo** [comment]
That makes sense, and is analogous to the Wine Conformance Test Suite - in that case they are measuring the behaviour of Windows vs Wine to check for conformance, people can run the test suite.

This is useful, as you might find there are actually slight differences in output - between different versions of MacOS and different generations of the hardware and drivers etc (Metal is on Intel and AMD based graphics, so there are bound to be small differences between them).

**Comment by steelbrain** [comment]
I've gotten pretty far with the metal2vulkan project (air2spir-v is just not a catchy name). Running it on every single metallib I could capture from macOS and installed apps, it is 97.31% byte-exact between macOS M3-Max and NVIDIA RTX 5080 Mobile. The other 2.69% is floating-point tolerances (hardware differences) and the like.

So I tasked an LLM to integrate it with Darling. It used https://github.com/darlinghq/indium as the base (but replaced its vulkan emitter with metal2vulkan), implemented some missing APIs, fixed some compatibility things with XWayland and got it working.

I'm not sure what to do with it since my main goal with metal2vulkan is different (QEMU ParaVirt GPU, host-side that runs on any vulkan based GPU). So if there's anyone interested from the darling side in working with me to pick this up and implement it proper (no/less slop), please write to me!

Here's a video demo of a metal shader running under a modified Darling and a modified Indium

https://github.com/user-attachments/assets/9d4161ee-3b22-4a18-a822-d83b857d862c

<details>
<summary>LLM Summary of its work</summary>

```
## Native macOS Metal app running on Linux — shaders compiled by Apple's own compiler, executed on the GPU via Vulkan, inside a real Cocoa window

We just got an end-to-end pipeline working : a **native macOS application** —
standard `Cocoa` + `Metal` + `MetalKit` API code, built as a Mach-O binary —
running under **Darling on Linux**, with its Metal shaders **compiled from source by Apple's actual Metal compiler**,
executed on the **Linux GPU through Vulkan**, and drawn in a **real native AppKit window with working native controls**.

### The demo
A real-time **raymarched Mandelbulb** fractal rendered in an embedded `MTKView`,
driven live by **native AppKit controls** — `NSSlider`s for the fractal exponent,
zoom, orbit angle and tilt, and an `NSButton` to cycle color palettes. Dragging a
slider morphs/rotates the fractal in real time; the sliders, button and the value
readout are genuine `NSView`/`NSControl` objects with real hit-testing and event
routing, not pixels painted into the Metal surface.

### The pipeline (every hop is real)
Shaders.metal  (Metal Shading Language source)
    │  Apple metal / metallib  ← Xcode 13.4.1 toolchain, running under Darling on Linux
    ▼
.metallib  (AIR bitcode)
    │  loaded at runtime by indium (Metal implementation)
    │  metal2vulkan:  AIR → SPIR-V
    ▼
SPIR-V  →  executed on the GPU via Vulkan
    │  rendered frame read back to the CPU
    ▼
XPutImage into the X11 window
In parallel, the surrounding native UI (`NSWindow` / `NSView` / `NSButton` /
`NSSlider`, drawn by cocotron's CoreGraphics) is presented to the same window and
routes real mouse events back into the AppKit responder chain.

### Why this is notable
- Apple's **real Metal shader compiler** runs on Linux (under Darling) and
  produces the `.metallib` the app loads — no reimplemented compiler.
- Arbitrary Metal Shading Language (a heavy per-pixel raymarcher with loops,
  SDF distance estimation, normals, AO, lighting) is translated **AIR → SPIR-V**
  and runs correctly on a commodity Linux GPU.
- The Metal surface **and** a native Cocoa control panel coexist and are both
  interactive in one window, on a stock Linux/Wayland desktop.

### Honest caveats (this is a working proof of concept, not a finished product)
- **Presentation is CPU-readback + `XPutImage`.** The GL/EGL swap path doesn't
  display on this host (Xwayland + NVIDIA), so frames are read back and blitted
  rather than zero-copy presented.
- **No depth buffer or texture sampling yet** in the Metal implementation — the
  demo is a raymarcher precisely because it needs neither.
- **Rendering is event-driven** (re-render per control change): the cocotron run
  loop doesn't reliably sustain a frame timer on this build, so continuous
  auto-animation isn't wired up yet.

Tested on a stock Linux desktop (Wayland/Xwayland, NVIDIA GPU). Built entirely
from source; example app under `examples/metal-controls`.
```
</details>

**Comment by stuaxo** [comment]
What do the commits look like ?

When I tried some LLM run dev for a PR on a much smaller project (System7 by Kelsie Davies) the first lot of commits weren't great.

What I got it do though afterwards was (after saving the current branch) and get it to build a branch per feature, give it feedback on those and then get it to squash each one into one commit, just like it would work as a PR.

This left the history in a much better state, to use as a refererence + I could get it to remove obviously LLMisms (the comments that don't add much, get it to add references to documentation etc - which I could then check and verify obvious things it got wrong).

**Comment by steelbrain** [comment]
metal2vulkan is shared with my paravirt-gpu project so the commit history is a mess, but its a rust binary that does one job really well so it might not matter how messy it is so long as it works?

Either way I'm looking to open source this in a month or so, at least the metal2vulkan part, then someone smarter than me can integrate it into Darling. Darling is missing a substantial amount of API surface for any modern app. Also needs modernization away from X to wayland, and away from OpenGL to Vulkan. Lots of tech debt but Its solvable

**Comment by steelbrain** [comment]
metal2vulkan is now open source: https://github.com/steelbrain/metal2vulkan

I’ll leave the rest to the community 🫡
